File: work/t1/BaseDataT1/BaseDataT1/BaseData.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Proto;

namespace CommonData
{
    public class BaseData
    {
        Object mValue = null;
        Dictionary<Object, BaseData> mObject = null;
        bool mIsBaseObject = false;
        public BaseData(Object value)
        {
            if (value is BaseData)
            {
                BaseData tempObject = (BaseData)value;
                if (tempObject.mIsBaseObject == true)
                {
                    this.mObject = tempObject.mObject;
                    this.mIsBaseObject = true;
                }
                else
                {
                    this.mValue = tempObject.mValue;
                }
            }
            else
            {
                if (value == null)
                {
                    mValue = null;
                }
                else if (value is int || value is float || value is double
                      || value is String || value is Boolean || value is long
                      || value is short)
                {
                    mValue = value;
                }
                else
                {
                    mValue = null;
                    Console.Write("HCData function HCData, value type error.");
                }
            }
        }
        public BaseData createObject()
        {
            mIsBaseObject = true;
            mObject = new Dictionary<Object, BaseData>();
            return this;
        }
        public BaseData putObject(Object key, Object value)
        {
            if (mIsBaseObject == true && (key is int || key is String))
            {
                if (value is BaseData)
                {
                    if (!(((BaseData)value).isObject() == false && ((BaseData)value).mValue == null))
                        mObject.Add(key, (BaseData)value);
                    else
                        mObject.Remove(key);
                }
                else
                {
                    if (value != null)
                        mObject.Add(key, new BaseData(value));
                    else
                        mObject.Remove(key);
                }
                return this;
            }
            else
            {
                Console.Write("HCData function putObject, is not object or key type error.");
                throw new Exception();
            }
        }
        public BaseData getObject(Object key)
        {
            if (mIsBaseObject == true)
            {
                BaseData bsData = mObject[key];
                if (bsData == null)
                {
                    return new BaseData(null);
                }
                return bsData;
            }
            else
            {
                Console.Write("HCData function getObject, is not object");
                throw new Exception();
            }
        }
        public bool isObject()
        {
            return mIsBaseObject;
        }
        public Object getBaseValue()
        {
            if (mIsBaseObject == false)
                return mValue;
            else
            {
                Console.Write("HCData function getBaseValue, is object");
                return null;
            }
        }
        public BaseData setBaseValue(Object value)
        {
            if (mIsBaseObject == false && (value is int || value is float || value is double
                    || value is String || value is bool || value is long
                    || value is short))
                return this;
            else
            {
                Console.WriteLine("HCData function getBaseValue, is object");
                throw new Exception();
            }
        }
        public override String ToString()
        {
            String returnString = "";
            if (mIsBaseObject == false)
            {
                if (mValue == null)
                    returnString = "null";
                else
                    returnString = mValue + "";
            }
            else
            {
                returnString += "{";
                String tempString = "";
                foreach (var item in mObject)
                {
                    returnString += tempString;
                    returnString += item.Key + "";
                    returnString += ":";
                    BaseData value = item.Value;
                    if (value.mIsBaseObject == false)
                    {
                        if (value.mValue == null)
                            returnString += "null";
                        else
                            returnString += value.mValue + "";
                    }
                    else
                    {
                        returnString += value.ToString();
                    }
                    tempString = ",";
                }
                returnString += "}";
            }
            return returnString;
        }

        public Proto.BaseData toProtoBuf()
        {
            Proto.BaseData bsData = new Proto.BaseData { };
            if (mIsBaseObject == false)
            {
                bsData.MIsBaseObject = false;

                if (mValue == null)
                {
                    bsData.MValue = new Proto.BaseData.Types.BaseValue { MType = 0 };
                    return bsData;
                }
                else
                {
                    if (mValue is int)
                    {
                        bsData.MValue = new Proto.BaseData.Types.BaseValue { BvInt = (int)mValue, MType = 1 };
                    }
                    else if (mValue is float)
                    {
                        bsData.MValue = new Proto.BaseData.Types.BaseValue { BvFloat = (float)mValue, MType = 2 };
                    }
                    else if (mValue is double)
                    {
                        bsData.MValue = new Proto.BaseData.Types.BaseValue { BvDouble = (double)mValue, MType = 3 };
                    }
                    else if (mValue is String)
                    {
                        bsData.MValue = new Proto.BaseData.Types.BaseValue { BvString = (String)mValue, MType = 4 };
                    }
                    else if (mValue is bool)
                    {
                        bsData.MValue = new Proto.BaseData.Types.BaseValue { BvBool = (bool)mValue, MType = 5 };
                    }
                    else if (mValue is long)
                    {
                        bsData.MValue = new Proto.BaseData.Types.BaseValue { BvLong = (long)mValue, MType = 6 };
                    }
                    else if (mValue is short)
                    {
                        bsData.MValue = new Proto.BaseData.Types.BaseValue { BvInt = (int)mValue, MType = 1 };
                    }
                    return bsData;
                }
            }
            else
            {
                bsData.MIsBaseObject = true;
                foreach (var item in mObject)
                {
                    BaseData value = item.Value;
                    Object key = item.Key;
                    Proto.BaseData.Types.BaseValue keyData = new Proto.BaseData.Types.BaseValue { };
                    if (key is String)
                    {
                        keyData.BvString = ((String)key);
                        keyData.MType = 4;
                    }
                    else if (key is int)
                    {
                        keyData.BvInt = ((int)key);
                        keyData.MType = 1;
                    }
                    else
                        continue;
                    Proto.BaseData.Types.ObjectType objData = new Proto.BaseData.Types.ObjectType { };
                    objData.MKey = keyData;
                    objData.MValue = value.toProtoBuf();
                    bsData.MObject.Add(objData);
                }
            }
            return bsData;
        }

        public static BaseData bytetoBaseData(Proto.BaseData proBD)
        {
            BaseData bsData = null;
            if (proBD.MIsBaseObject != true)
            {
                if (proBD.MValue == null)
                {
                    bsData = new BaseData(null);
                }
                else if (proBD.MValue.MType == 0)
                {
                    bsData = new BaseData(null);
                }
                else if (proBD.MValue.MType == 1)
                {
                    bsData = new BaseData(proBD.MValue.BvInt);
                }
                else if (proBD.MValue.MType == 2)
                {
                    bsData = new BaseData(proBD.MValue.BvFloat);
                }
                else if (proBD.MValue.MType == 3)
                {
                    bsData = new BaseData(proBD.MValue.BvDouble);
                }
                else if (proBD.MValue.MType == 4)
                {
                    bsData = new BaseData(proBD.MValue.BvString);
                }
                else if (proBD.MValue.MType == 5)
                {
                    bsData = new BaseData(proBD.MValue.BvBool);
                }
                else if (proBD.MValue.MType == 6)
                {
                    bsData = new BaseData(proBD.MValue.BvLong);
                }
            }
            else
            {
                bsData = new BaseData(null);
                bsData.createObject();
                int objSize = proBD.MObject.Count;
                for (int i = 0; i < objSize; i++)
                {
                    Proto.BaseData.Types.ObjectType keyValue = proBD.MObject[i];
                    if (keyValue.MKey.MType == 1)
                    {
                        bsData.putObject(keyValue.MKey.BvInt, bytetoBaseData(keyValue.MValue));
                    }
                    else if (keyValue.MKey.MType == 4)
                    {
                        bsData.putObject(keyValue.MKey.BvString, bytetoBaseData(keyValue.MValue));
                    }
                }
            }
            return bsData;
        }

    }
}

File: work/HCP/src/main/java/com/hcp/util/BaseData.java
package com.hcp.util;

import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import org.luaj.vm2.LuaTable;
import org.luaj.vm2.LuaValue;

import com.hcp.proto.BaseDataProtos;

public class BaseData {
	Object mValue = null;
	Map<Object, BaseData> mObject = null;
	boolean mIsBaseObject = false;

	public BaseData(Object value) throws Exception {
		if (value instanceof BaseData) {
			BaseData tempObject = (BaseData) value;
			if (tempObject.mIsBaseObject == true) {
				this.mObject = tempObject.mObject;
				this.mIsBaseObject = true;
			} else {
				this.mValue = tempObject.mValue;
			}
		} else {
			if (value == null) {
				mValue = null;
			} else if (value instanceof Integer || value instanceof Float || value instanceof Double
					|| value instanceof String || value instanceof Boolean || value instanceof Long
					|| value instanceof Short) {
				mValue = value;
			} else {
				mValue = null;
				System.out.println("BaseData function BaseData, arg type error.");
				throw new Exception();
			}
		}
	}

	public BaseData createObject() {
		mIsBaseObject = true;
		mObject = new ConcurrentHashMap<Object, BaseData>();
		return this;
	}

	public Iterator<Map.Entry<Object, BaseData>> iterator() {
		if (mIsBaseObject == false)
			return null;
		return mObject.entrySet().iterator();
	}

	public BaseData putObject(Object key, Object value) throws Exception {
		if (mIsBaseObject == true && (key instanceof Integer || key instanceof String)) {
			if (value instanceof BaseData) {
				if (!(((BaseData) value).mIsBaseObject == false && ((BaseData) value).mValue == null))
					mObject.put(key, (BaseData) value);
				else
					mObject.remove(key);
			} else {
				if (value != null)
					mObject.put(key, new BaseData(value));
				else
					mObject.remove(key);
			}
			return this;
		} else {
			System.out.println("BaseData function putObject, is not object or key type error.");
			throw new Exception();
		}
	}

	/**
	 * this function is so low
	 * 
	 * @throws Exception
	 */
	public BaseData getObject(Object key) throws Exception {
		if (mIsBaseObject == true) {
			BaseData bsData = mObject.get(key);
			if (bsData == null) {
				return new BaseData(null);
			}
			return bsData;
		} else {
			System.out.println("BaseData function getObject, is not object");
			throw new Exception();
		}
	}

	public boolean isObject() {
		return mIsBaseObject;
	}

	public Object getBaseValue() throws Exception {
		if (mIsBaseObject == false)
			return mValue;
		else {
			System.out.println("BaseData function getBaseValue, is object");
			throw new Exception();
		}
	}

	public BaseData setBaseValue(Object value) throws Exception {
		if (mIsBaseObject == false && (value instanceof Integer || value instanceof Float || value instanceof Double
				|| value instanceof String || value instanceof Boolean || value instanceof Long
				|| value instanceof Short))
			return this;
		else {
			System.out.println("BaseData function setBaseValue, is object");
			throw new Exception();
		}
	}

	public LuaValue toLuaValue() {
		if (this.mIsBaseObject == false) {
			if (mValue == null) {
				return LuaValue.NIL;
			} else {
				if (mValue instanceof Integer) {
					return LuaValue.valueOf((int) mValue);
				} else if (mValue instanceof Float) {
					return LuaValue.valueOf((float) mValue);
				} else if (mValue instanceof Double) {
					return LuaValue.valueOf((double) mValue);
				} else if (mValue instanceof String) {
					return LuaValue.valueOf((String) mValue);
				} else if (mValue instanceof Boolean) {
					return LuaValue.valueOf((boolean) mValue);
				} else if (mValue instanceof Long) {
					return LuaValue.valueOf((long) mValue);
				} else if (mValue instanceof Short) {
					return LuaValue.valueOf((Short) mValue);
				}
			}
		} else {
			LuaValue table = new LuaTable();
			for (Map.Entry<Object, BaseData> entry : mObject.entrySet()) {
				BaseData value = entry.getValue();
				Object key = entry.getKey();
				if (key instanceof String) {
					table.set((String) key, value.toLuaValue());
				} else if (key instanceof Integer) {
					table.set((int) key, value.toLuaValue());
				} else
					continue;
			}
			return table;
		}
		return null;
	}

	@SuppressWarnings("unused")
	public static BaseData parseFromLuaValue(LuaValue lv) throws Exception {
		BaseData bsData = null;
		if (lv.istable() == false) {
			if (lv == null) {
				bsData = new BaseData(null);
			} else if (lv.isnil()) {
				bsData = new BaseData(null);
			} else if (lv.isstring()) {
				bsData = new BaseData(lv.tojstring());
			} else if (lv.isint()) {
				bsData = new BaseData(lv.toint());
			} else if (lv.islong()) {
				bsData = new BaseData(lv.tolong());
			} else if (false) {
				bsData = new BaseData(null);
			} else if (lv.isnumber()) {
				bsData = new BaseData(lv.todouble());
			} else if (lv.isboolean()) {
				bsData = new BaseData(lv.toboolean());
			}
		} else {
			bsData = new BaseData(null);
			bsData.createObject();
			LuaTable t = (LuaTable) lv;
			LuaValue[] keys = t.keys();
			for (int i = 0; i < keys.length; i++) {
				LuaValue k = keys[i];
				if (k.isint()) {
					bsData.putObject(k.toint(), parseFromLuaValue(t.get(k)));
				} else if (k.isstring()) {
					bsData.putObject(k.tojstring(), parseFromLuaValue(t.get(k)));
				}
			}
		}
		return bsData;
	}

	public BaseDataProtos.BaseData toProtoBuf() throws Exception {
		BaseDataProtos.BaseData.Builder bsBuilder = BaseDataProtos.BaseData.newBuilder();
		if (mIsBaseObject == false) {
			bsBuilder.setMIsBaseObject(false);
			if (mValue == null) {
				BaseDataProtos.BaseData.BaseValue.Builder bsvBuilder = BaseDataProtos.BaseData.BaseValue.newBuilder();
				bsvBuilder.setMType(0);
				bsBuilder.setMValue(bsvBuilder.build());
				return bsBuilder.build();
			} else {
				BaseDataProtos.BaseData.BaseValue.Builder bsvBuilder = BaseDataProtos.BaseData.BaseValue.newBuilder();
				if (mValue instanceof Integer) {
					bsvBuilder.setBvInt((int) mValue);
					bsvBuilder.setMType(1);
				} else if (mValue instanceof Float) {
					bsvBuilder.setBvFloat((float) mValue);
					bsvBuilder.setMType(2);
				} else if (mValue instanceof Double) {
					bsvBuilder.setBvDouble((double) mValue);
					bsvBuilder.setMType(3);
				} else if (mValue instanceof String) {
					bsvBuilder.setBvString((String) mValue);
					bsvBuilder.setMType(4);
				} else if (mValue instanceof Boolean) {
					bsvBuilder.setBvBool((boolean) mValue);
					bsvBuilder.setMType(5);
				} else if (mValue instanceof Long) {
					bsvBuilder.setBvLong((long) mValue);
					bsvBuilder.setMType(6);
				} else if (mValue instanceof Short) {
					Integer value = ((Short) mValue).intValue();
					bsvBuilder.setBvInt(value);
					bsvBuilder.setMType(1);
				}
				bsBuilder.setMValue(bsvBuilder.build());
				return bsBuilder.build();
			}
		} else {
			bsBuilder.setMIsBaseObject(true);
			for (Map.Entry<Object, BaseData> entry : mObject.entrySet()) {
				BaseData value = entry.getValue();
				Object key = entry.getKey();
				BaseDataProtos.BaseData.BaseValue.Builder keyBuilder = BaseDataProtos.BaseData.BaseValue.newBuilder();
				if (key instanceof String) {
					keyBuilder.setMType(4);
					keyBuilder.setBvString((String) key);

				} else if (key instanceof Integer) {
					keyBuilder.setMType(1);
					keyBuilder.setBvInt((int) key);
				} else
					continue;
				BaseDataProtos.BaseData.ObjectType.Builder objBuilder = BaseDataProtos.BaseData.ObjectType.newBuilder();
				objBuilder.setMKey(keyBuilder.build());
				objBuilder.setMValue(value.toProtoBuf());
				bsBuilder.addMObject(objBuilder.build());
			}
		}
		return bsBuilder.build();
	}

	public static BaseData bytetoBaseData(BaseDataProtos.BaseData proBD) throws Exception {
		BaseData bsData = null;
		if (proBD.getMIsBaseObject() == false) {
			if (proBD.getMValue() == null) {
				bsData = new BaseData(null);
			} else if (proBD.getMValue().getMType() == 0) {
				bsData = new BaseData(null);
			} else if (proBD.getMValue().getMType() == 1) {
				bsData = new BaseData(proBD.getMValue().getBvInt());
			} else if (proBD.getMValue().getMType() == 2) {
				bsData = new BaseData(proBD.getMValue().getBvFloat());
			} else if (proBD.getMValue().getMType() == 3) {
				bsData = new BaseData(proBD.getMValue().getBvDouble());
			} else if (proBD.getMValue().getMType() == 4) {
				bsData = new BaseData(proBD.getMValue().getBvString());
			} else if (proBD.getMValue().getMType() == 5) {
				bsData = new BaseData(proBD.getMValue().getBvBool());
			} else if (proBD.getMValue().getMType() == 6) {
				bsData = new BaseData(proBD.getMValue().getBvLong());
			}
		} else {
			bsData = new BaseData(null);
			bsData.createObject();
			Iterator<BaseDataProtos.BaseData.ObjectType> it = proBD.getMObjectList().iterator();
			while (it.hasNext()) {
				BaseDataProtos.BaseData.ObjectType keyValue = it.next();
				if (keyValue.getMKey().getMType() == 1) {
					bsData.putObject(keyValue.getMKey().getBvInt(), bytetoBaseData(keyValue.getMValue()));
				} else if (keyValue.getMKey().getMType() == 4) {
					bsData.putObject(keyValue.getMKey().getBvString(), bytetoBaseData(keyValue.getMValue()));
				}
			}
		}
		return bsData;
	}

	public String toString() {
		String returnString = "";
		if (mIsBaseObject == false) {
			if (mValue == null)
				returnString = "null";
			else
				returnString = mValue.toString();
		} else {
			returnString += "{";
			String tempString = "";
			for (Map.Entry<Object, BaseData> entry : mObject.entrySet()) {
				returnString += tempString;
				returnString += entry.getKey().toString();
				returnString += ":";
				BaseData value = entry.getValue();
				if (value.mIsBaseObject == false) {
					if (value.mValue == null)
						returnString += "null";
					else
						returnString += value.mValue.toString();
				} else {
					returnString += value.toString();
				}
				tempString = ",";
			}
			returnString += "}";
		}
		return returnString;
	}
}
